### ISSUE TITLE
kv: a batch with multiple ReverseScan requests fails

### ISSUE THREAD

**Issue by vivekmenezes** [cockroachdb/cockroach#8245]
I tried adding a test with a batch with multiple ReverseScan requests in #7725 and it had surprising results. One option is to limit the number of ReverseScan requests to 1 if SQL doesn't care about support for many ReverseScan requests in a single batch.

**Comment by petermattis** [comment]
Can you add details of what fails with multiple ReverseScan requests in a single batch? SQL execution definitely expects this to work.

**Comment by bdarnell** [comment]
@petermattis Why does SQL need multiple ReverseScans in the same batch? In #7725 I was suggesting that when a limit is used (and I assume all SQL scans will eventually use limits if they don't already), it would be much simpler to only allow one request per batch when there is a limit instead of applying a single limit across all requests in the batch. You already have to separate Scans from ReverseScans since you can't mix directions in a single batch.

**Comment by petermattis** [comment]
Index selection can generate multiple scan requests for a query like `SELECT * FROM a in (1, 3, 5, 7, 9)`. If you add an `ORDER BY DESC` clause, index selection will generate reverse scans (depending on your index).

**Comment by petermattis** [comment]
Cc @RaduBerinde who is more familiar with the index selection code now than I am.

**Comment by bdarnell** [comment]
And how important is it that these all go in the same batch?

**Comment by petermattis** [comment]
Fairly important. Otherwise you're doing a round trip for each scan.

**Comment by RaduBerinde** [comment]
The use of Scan and ReverseScan is fairly symmetric in the SQL layer, we pretty much need the same semantics for both.

In addition to `ORDER BY DESC LIMIT xx`, another example is `SELECT MAX(x) FROM t` with a WHERE clause. We only want one row but we may be looking in many spans, any of which could have no keys.

What is the failure? Does it show up on master? I am surprised our tests didn't catch it..

**Comment by vivekmenezes** [comment]
This is actually working. I've created a PR with a test for concurrent ReverseScans. This also includes a more accurate test for concurrent Scans.

**Comment by vivekmenezes** [comment]
Reopening this issue because I wanted to discuss another issue. The way we implement the range operations now: we iterate through all the ranges one by one in a sorted order and apply all the requests on each range. The result is that if there is an overlap in the key space of requests, when the batch request bound is met, we can end up with multiple responses that are incomplete.

with the database containing keys "a1", "a2", "b1", in ranges ["a"-"b"] and ["b"-"c"]

commands
batch.Scan("a", "c")
batch.Scan("a", "c")

with a bound of say 3 will return results

"a1", "a2"
"a1"

and not

"a1", "a2", "b1"
and 
an empty response

This is probably not a problem because I suspect SQL isn't going to issue commands in a batch with overlapping keys. I'd like to know whether we should just disallow overlapping keys.

**Comment by RaduBerinde** [comment]
Disallowing overlapping spans is ok from the SQL side, and I agree it would make sense semantically.

**Comment by petermattis** [comment]
Checking that spans overlap might be expensive, though.

**Comment by RaduBerinde** [comment]
From the SQL side it's also ok to require the spans in a request to be sorted (increasing for `Scans`, decreasing for `ReverseScan`s) if that helps.

**Comment by RaduBerinde** [comment]
Ah, actually scratch that. They might not be when we do an index join (and later a join-by-lookup)..

**Comment by vivekmenezes** [comment]
Actually the semantics are still going to be messed up I believe if the
requests are not ordered. Let me write a test to verify its indeed the case.

**Comment by RaduBerinde** [comment]
Note that in those cases each range corresponds to a single SQL row and we would only be doing Scans (no ReverseScans).

**Comment by vivekmenezes** [comment]
I'd further add that batches as described in the comments can have their
requests processed in parallel, so returning partial responses to many
requests is valid. Question is is that incompatible with SQL needs.

**Comment by RaduBerinde** [comment]
Right now any kind of partial response is incompatible with SQL.

Whether we can change that depends what "partial responses" means exactly and if the SQL layer can figure out it was a partial response or not.

Generally, having partial responses for multiple spans is problematic as the only thing SQL can do is keep all the partial responses in memory, adjust the spans and try again.. If we keep getting partial results memory usage will keep increasing.

**Comment by vivekmenezes** [comment]
I think for now we can go with the model where the kv layer is allowed to return partial responses. Each partial response will contain a ResumeSpan, which will allow SQL to resume a request in its next call to kv with a fresh updated Span. I suspect that in the future we need to define a different API here based on more requirements from dist SQL. I think the best way to do that is to use the current one see where the memory buffering that you mention happens and fix those kinds of issues either by splitting the pathological batch within SQL, or defining a new behavior for the batch supported in kv. Let's do that later.

**Comment by RaduBerinde** [comment]
So the issue is this. SQL might need to output all the rows within spans `[A-B), [D-F), [O-T)`, **in order**. If we get partial responses for all these spans, we can't output anything inside `[D-F)` or `[O-T)` until we're sure we got everything in `[A-B)`. So we would need to issue more requests and we either have to remember all the partial responses we got for those later spans, or discard the results and try again (neither is ideal, unless this is something that is guaranteed to happen rarely).

Also, I don't really understand how this proposed interface would work with respect to batch limits. Would the limit be "advisory" and we may receive more results? (can it be significantly more?) Note that SQL almost always sets a batch limit (unless it knows there's only one row per span).

**Comment by vivekmenezes** [comment]
BTW the proposed interface is how it works today (without ResumeSpan which I'm adding #7725)

If you do send a request that is ordered ```[A-B), [D-F), [O-T)``` you will not receive keys for D and beyond . The implementation delivers a response with multiple partial responses when you have overlapping keys like
```[A-D), [B-F), [O-T)``` (for which you might see C show up for both response 0 and 1) or non-overlapping keys but unsorted order like ```[B-F), [A-B), [O-T)``` where you might see more than one partial response (See the new test TestMultiRangeBoundedBatchScanUnsortedOrder I just added to 
#8268 )

**Comment by vivekmenezes** [comment]
TestMultiRangeBoundedBatchScanUnsortedOrder

and

TestMultiRangeBoundedBatchScanSortedOverlapping

illustrate the two cases

**Comment by RaduBerinde** [comment]
For the unsorted non-overlapping case - can that happen even if there is no limit (MaxSpanRequestKeys) on the request? Or are we guaranteed full results if MaxSpanRequestKeys isn't set?

SQL cares about that case for index joins, but it doesn't have to set MaxSpanRequestKeys in that case.

**Comment by vivekmenezes** [comment]
We've fixed all the outstanding issues here. I'm closing this issue.
